### DISCUSSION TITLE
Not reachable envoy endpoints outside of container in headless services

### DISCUSSION THREAD

**Discussion by ckaipf** [istio/istio#52789] [Q&A] [answered]
Hi, 

I'm looking for support to debug not reachable envoy endpoints.

We use a Bitnami chart installed MongoDB (replica set and TLS enabled) in our K8s cluster:

```k get po
NAME        READY   STATUS    RESTARTS   AGE
busybox     1/1     Running   0          4h28m
mongodb-0   2/2     Running   0          81m
mongodb-1   2/2     Running   0          81m
mongodb-2   2/2     Running   0          83m
```

```
 k get svc
NAME               TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)     AGE
mongodb-headless   ClusterIP   None         <none>        27017/TCP   5d
```

``` 
istioctl proxy-status
NAME                                                    CLUSTER        CDS        LDS        EDS        RDS        ECDS     ISTIOD                      VERSION
mongodb-0.database                                      Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED              istiod-5cbbfb6b4c-4fdqv     1.23.0
mongodb-1.database                                      Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED              istiod-5cbbfb6b4c-4fdqv     1.23.0
mongodb-2.database                                      Kubernetes     SYNCED     SYNCED     SYNCED     SYNCED              istiod-5cbbfb6b4c-4fdqv     1.23.0
```

Within the container I'm able to curl e.g. the Prometheus stats endpoint:

```
istio-proxy@mongodb-0:/$ curl -s http://mongodb-headless:15090/stats/prometheus | head
# TYPE envoy_cluster_assignment_stale counter
envoy_cluster_assignment_stale{cluster_name="xds-grpc"} 0
# TYPE envoy_cluster_assignment_timeout_received counter
envoy_cluster_assignment_timeout_received{cluster_name="xds-grpc"} 0
# TYPE envoy_cluster_assignment_use_cached counter
envoy_cluster_assignment_use_cached{cluster_name="xds-grpc"} 0
# TYPE envoy_cluster_bind_errors counter
envoy_cluster_bind_errors{cluster_name="xds-grpc"} 0
# TYPE envoy_cluster_default_total_match_count counter
envoy_cluster_default_total_match_count{cluster_name="xds-grpc"} 1
```

However, outside of it, it times out:

```
busybox:~# curl http://mongodb-headless:15090/stats/prometheus -v
* Host mongodb-headless:15090 was resolved.
* IPv6: (none)
* IPv4: 172.25.2.40, 172.25.1.217, 172.25.0.184
*   Trying 172.25.2.40:15090...
```

The same applies to our Strimzi operated Kafka. I'm not sure if this is related to the applications TLS or the headless service. Does anyone has an idea how to debug this?

Thanks!

**Comment by howardjohn** [comment]
This call is from <pod not in mesh> --> <pod in the mesh on a port that is not captured> (15090 is excluded from IPTables redirection), so there should be no Istio on the path at all...

**Reply by ckaipf** [reply]
Thanks for your answer! 

Right, because those are endpoints exposed by the envoy proxy, or? 

The problem behind what I'm trying to solve is that Prometheus can't scrape those endpoints (time out/target down). However, for other services it works as expected.

This is an example from Prometheus targets (envoy-stats), targets `down`. Targets are down for Kafka and MongoDB. Both use a headless service. 

```
Endpoint | State | Labels | Last Scrape | Scrape Duration | Error
http://172.25.1.19:15090/stats/prometheus | UP | instance="172.25.1.19:15090"job="envoy-stats" | 5.969s ago | 7.298ms |  
http://172.25.2.11:15090/stats/prometheus | DOWN | instance="172.25.2.11:15090"job="envoy-stats" | 12.586s ago | 10.0s | Get "http://172.25.2.11:15090/stats/prometheus": context deadline exceeded
```

So I'm wondering why it is possible to scrape some endpoints and some not.

**Reply by ckaipf** [reply, marked as answer]
I could solve this by injecting a istio sidecar in the clients. Both the busybox in the example and the Prometheus pod didn't had one. Since I'm new to istio and for curiosity do you @howardjohn know why some envoy metrics can be scraped without having an injected sidecar in the client and some not?
